File: Striver/DP/Q34.java
package Striver.DP;

public class Q34{
    // wildcard string matching
    // ? can be replaced by any single character
    // * can be replaced by any sequence of characters or empty sequence

    // recursive solution - Time: O(2^n) Space: O(n)
    public boolean isMatch(String s, String p, int i, int j){
        if(i < 0 && j < 0) return true;
        if(i < 0 && j >= 0) return false;
        if(j < 0 && i >= 0) {
            for(int k = 0; k <= i; k++){
                if(s.charAt(k) != '*') return false;
            }
            return true;
        }

        if(s.charAt(i) == p.charAt(j) || s.charAt(i) == '?') return isMatch(s, p, i-1, j-1);
        if(s.charAt(i) == '*'){
            return isMatch(s, p, i-1, j) || isMatch(s, p, i, j-1);
        }

        return false;
    }

    // memoization - Time: O(n^2) Space: O(n^2)
    public boolean isMatchMem(String s, String p, int i, int j, int[][] dp){
        if(i < 0 && j < 0) return true;
        if(i < 0 && j >= 0) return false;
        if(j < 0 && i >= 0) {
            for(int k = 0; k <= i; k++){
                if(s.charAt(k) != '*') return false;
            }
            return true;
        }

        if(dp[i][j] != -1) return dp[i][j] == 1;

        if(s.charAt(i) == p.charAt(j) || s.charAt(i) == '?'){
            
            dp[i][j] = isMatchMem(s, p, i-1, j-1, dp) ? 1 : 0;
            return dp[i][j] == 1;
        } 
        if(s.charAt(i) == '*'){
            dp[i][j] = (isMatchMem(s, p, i-1, j, dp) || isMatchMem(s, p, i, j-1, dp)) ? 1 : 0;
            return dp[i][j] == 1;
        }

        dp[i][j] = 0;
        return false;
    }

    // tabulation - Time: O(n^2) Space: O(n^2)
    public boolean isMatchTab(String s, String p){
        int n = s.length(), m = p.length();
        boolean[][] dp = new boolean[n+1][m+1];
        dp[0][0] = true;
        for(int j = 1; j <=m; j++){
            dp[0][j] = false;
        }

        for(int i = 1; i <= n; i++){
            boolean flag = true;
            for(int k = 1; k <= i; k++){
                if(s.charAt(k-1) != '*'){
                    flag = false;
                    break;
                }
            }
            dp[i][0] = flag;
        }

        for(int i = 1; i <= n; i++){
            for(int j = 1; j <= m; j++){
                if(s.charAt(i-1) == p.charAt(j-1) || s.charAt(i-1) == '?') dp[i][j] = dp[i-1][j-1];
                else if(s.charAt(i-1) == '*') dp[i][j] = dp[i-1][j] || dp[i][j-1];
                else dp[i][j] = false;
            }
        }

        return dp[n][m];
    }

    // space optimization - Time: O(n^2) Space: O(n)
    public boolean isMatchOpt(String s, String p){
        int n = s.length(), m = p.length();
        boolean[] dp = new boolean[m+1];
        boolean[] curr = new boolean[m+1];
        dp[0] = true;
        for(int j = 1; j <=m; j++){
            dp[j] = false;
        }

        for(int i = 1; i <= n; i++){
            boolean flag = true;
            for(int k = 1; k <= i; k++){
                if(s.charAt(k-1) == '*'){
                    flag = false;
                    break;
                }
            }
            curr[0] = flag;
            for(int j = 1; j <= m; j++){
                if(s.charAt(i-1) == p.charAt(j-1) || s.charAt(i-1) == '?') curr[j] = dp[j-1];
                else if(s.charAt(i-1) == '*') curr[j] = dp[j] || curr[j-1];
                else curr[j] = false;
            }
            for(int j = 0; j <= m; j++){
                dp[j] = curr[j];
            }
        }

        return dp[m];
    }

    public static void main(String[] args) {
        Q34 q = new Q34();
        String s = "ba*****ab";
        String p = "baxnab";
        System.out.println(q.isMatch(s, p, s.length()-1, p.length()-1));
        int[][] dp = new int[s.length()][p.length()];
        for(int i = 0; i < s.length(); i++){
            for(int j = 0; j < p.length(); j++){
                dp[i][j] = -1;
            }
        }
        System.out.println(q.isMatchMem(s, p, s.length()-1, p.length()-1, dp));
        System.out.println(q.isMatchTab(s, p));
        System.out.println(q.isMatchOpt(s, p));
    }

}

File: Striver/DP/Q34.py
# wildcard string matching
# ? - matches with any single character
# * - matches with any sequence of characters or empty string

# s is the string with ? and *
# p is the normal string

#recursive solution
def isMatch(s, p, i, j):
    if i < 0 and j < 0:
        return True
    if i < 0 and j >= 0:
        return False
    if j < 0 and i >= 0:
        for k in range(i+1):
            if s[k] != '*':
                return False
        return True
    
    if s[i] == p[j] or s[i] == '?':
        return isMatch(s, p, i-1, j-1)
    
    if s[i] == '*':
        return isMatch(s, p, i-1, j) or isMatch(s, p, i, j-1)
    
    return False


# memoization solution
def isMatchMem(s, p, i, j, dp):
    if i < 0 and j < 0:
        return True
    if i < 0 and j >= 0:
        return False
    if j < 0 and i >= 0:
        for k in range(i+1):
            if s[k] != '*':
                return False
        return True
    
    if dp[i][j] != -1:
        return dp[i][j]
    
    if s[i] == p[j] or s[i] == '?':
        dp[i][j] = isMatchMem(s, p, i-1, j-1, dp)
        return dp[i][j]
    
    if s[i] == '*':
        dp[i][j] = isMatchMem(s, p, i-1, j, dp) or isMatchMem(s, p, i, j-1, dp)
        return dp[i][j]
    
    dp[i][j] = False
    return dp[i][j]


# tabulation solution
def isMatchTab(s, p):
    n = len(s)
    m = len(p)
    dp = [[False for i in range(m+1)] for j in range(n+1)]
    dp[0][0] = True

    for j in range(1,m+1):
        dp[0][j] = False

    for i in range(1,n+1):
        flag = True
        for k in range(1,i+1):
            if s[k-1] != '*':
                flag = False
                break
        dp[i][0] = flag

    for i in range(1,n+1):
        for j in range(1,m+1):
            if s[i-1] == p[j-1] or s[i-1] == '?':
                dp[i][j] = dp[i-1][j-1]
            elif s[i-1] == '*':
                dp[i][j] = dp[i-1][j] or dp[i][j-1]
            else:
                dp[i][j] = False

    return dp[n][m]


# space optimization solution
def isMatchSpace(s, p):
    n = len(s)
    m = len(p)
    dp = [False for i in range(m+1)]
    curr = [False for i in range(m+1)]
    dp[0] = True

    # for j in range(1,m+1):
    #     dp[j] = False


    for i in range(1,n+1):
        flag = True
        for k in range(1,i+1):
            if s[k-1] != '*':
                flag = False
                break
        curr[0] = flag
        for j in range(1,m+1):
            if s[i-1] == p[j-1] or s[i-1] == '?':
                curr[j] = dp[j-1]
            elif s[i-1] == '*':
                curr[j] = curr[j-1] or dp[j]
            else:
                curr[j] = False
        dp = curr.copy()

    return dp[m]


s = "b*c*"
p = "backing"

print(isMatch(s, p, len(s)-1, len(p)-1))
dp = [[-1 for i in range(len(p))] for j in range(len(s))]
print(isMatchMem(s, p, len(s)-1, len(p)-1, dp))
print(isMatchTab(s, p))
print(isMatchSpace(s, p))
